cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnBuildSideBatch(size_t thread_index, ExecBatch batch)
+        {
+            return build_accumulator_.InsertBatch(
+                thread_index,
+                std::move(batch));
+        }
+
+        Status SpillingHashJoin::OnBuildSideFinished(size_t thread_index)
+        {
+            callbacks_.resume_probe_side(backpressure_counter_.fetch_add(1));
+            return BuildBloomFilter(thread_index);
+        }
+
+        Status SpillingHashJoin::BuildBloomFilter(size_t thread_index)
+        {
+            // Disable Bloom filter if bloom_filter_ = nullptr by advancing to past
+            // the final Bloom filter
+            partition_idx_ = (bloom_filter_ == nullptr)
+                ? partition_idx_ = SpillingAccumulationQueue::kNumPartitions
+                : build_accumulator_.hash_cursor();
+            return BuildNextBloomFilter(thread_index);
+        }
+
+        Status SpillingHashJoin::PushBloomFilterBatch(size_t thread_index, int64_t batch_id)
+        {
+            const uint64_t *hashes = build_accumulator_.GetHashes(
+                partition_idx_,
+                static_cast<size_t>(batch_id));
+            size_t num_rows = build_accumulator_.row_count(
+                partition_idx_,
+                static_cast<size_t>(batch_id));
+            return builder_->PushNextBatch(
+                thread_index,
+                static_cast<int64_t>(num_rows),
+                hashes);
+        }
+
+        Status SpillingHashJoin::BuildNextBloomFilter(size_t thread_index)
+        {
+            if(partition_idx_ >= SpillingAccumulationQueue::kNumPartitions)
+                return OnBloomFilterFinished(thread_index);
+
+            size_t num_rows = build_accumulator_.row_count(partition_idx_);
+            size_t num_batches = build_accumulator_.batch_count(partition_idx_);
+
+            // partition_idx_ is incremented in the callback for the taskgroup
+            bloom_filter_->partitions[partition_idx_] =
+                arrow::internal::make_unique<BlockedBloomFilter>();
+
+            RETURN_NOT_OK(builder_->Begin(
+                              num_threads_,
+                              ctx_->cpu_info()->hardware_flags(),
+                              ctx_->memory_pool(),
+                              num_rows,
+                              num_batches,
+                              bloom_filter_->partitions[partition_idx_].get()));
+                
+            return callbacks_.start_task_group(
+                task_group_bloom_[partition_idx_],
+                build_accumulator_.batch_count(partition_idx_));
+        }
+        
+        Status SpillingHashJoin::OnBloomFilterFinished(size_t thread_index)
+        {
+            RETURN_NOT_OK(callbacks_.bloom_filter_finished(thread_index));
+            bool expected = false;
+            if(!bloom_or_probe_finished_.compare_exchange_strong(expected, true))
```

Comment:
why the name bloom_or_probe_finished_? Why are both events glued together as alternative? Can we just have bloom_finished_ and if probe side finishes first then ignore it and still wait for Bloom filter build to finish?

Reply:
don't we still have to wait for probe side to finish before beginning the join?